### ISSUE TITLE
sigma.exceptions.SigmaRegularExpressionError: Regular expression 'X' is invalid: bad escape (end of pattern) at position Y

### ISSUE THREAD

**Issue by Res260** [SigmaHQ/pySigma#249]
Hi, I stumbled upon a weird problem regarding `SigmaRule.to_dict()`. Here is the code snippet to reproduce the bug:

```python
import yaml

from sigma.rule import SigmaRule

yaml_rule = r"""
title: A
id: aaaaaaaa-aaaa-4aaa-aaaa-aaaaaaaaaaaa
status: experimental
level: medium
logsource:
  product: something
detection:
  selection:
    field2|re|i:
      - 'something\$\\'
  condition: selection
"""

yaml_parsed = yaml.safe_load(yaml_rule)

rule_original = SigmaRule.from_dict(yaml_parsed)

rule_exported_to_dict = rule_original.to_dict()

rule_imported_from_dict = SigmaRule.from_dict(rule_exported_to_dict)
"""

yaml_parsed = yaml.safe_load(yaml_rule)

rule_original = SigmaRule.from_dict(yaml_parsed)

rule_exported_to_dict = rule_original.to_dict()

rule_imported_from_dict = SigmaRule.from_dict(rule_exported_to_dict)
```

This code basically parse a YAML rule into a dict, create a `SigmaRule` with it (no error). However, when we take this `SigmaRule` and call `.to_dict()` and create another `SigmaRule` with it, a regex error is thrown, which seems to indicate that there is an issue with `SigmaRule.to_dict()`:

```
Traceback (most recent call last):
  File "C:\<redacted>\pySigma\sigma\types.py", line 627, in compile
    re.compile(self.regexp, flags)
  File "C:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\re\__init__.py", line 228, in compile
    return _compile(pattern, flags)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\re\__init__.py", line 307, in _compile
    p = _compiler.compile(pattern, flags)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\re\_compiler.py", line 745, in compile
    p = _parser.parse(p, flags)
        ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\re\_parser.py", line 979, in parse
    p = _parse_sub(source, state, flags & SRE_FLAG_VERBOSE, 0)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\re\_parser.py", line 460, in _parse_sub
    itemsappend(_parse(source, state, verbose, nested + 1,
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\re\_parser.py", line 530, in _parse
    sourceget()
  File "C:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\re\_parser.py", line 263, in get
    self.__next()
  File "C:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\re\_parser.py", line 252, in __next
    raise error("bad escape (end of pattern)",
re.error: bad escape (end of pattern) at position 11

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\<redacted>\AppData\Roaming\JetBrains\PyCharm2024.1\scratches\scratch_10.py", line 27, in <module>
    rule_imported_from_dict = SigmaRule.from_dict(rule_exported_to_dict)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\<redacted>\pySigma\sigma\rule.py", line 1177, in from_dict
    raise errors[0]
  File "C:\<redacted>\pySigma\sigma\rule.py", line 1166, in from_dict
    detections = SigmaDetections.from_dict(rule["detection"], source)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\<redacted>\pySigma\sigma\rule.py", line 703, in from_dict
    name: SigmaDetection.from_definition(definition, source)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\<redacted>\pySigma\sigma\rule.py", line 515, in from_definition
    SigmaDetectionItem.from_mapping(key, val, source)
  File "C:\<redacted>\pySigma\sigma\rule.py", line 374, in from_mapping
    return cls(field, modifiers, val, source=source)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<string>", line 9, in __init__
  File "C:\<redacted>\pySigma\sigma\rule.py", line 310, in __post_init__
    self.apply_modifiers()
  File "C:\<redacted>\pySigma\sigma\rule.py", line 322, in apply_modifiers
    self.value = [item for val in self.value for item in modifier_instance.apply(val)]
                                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\<redacted>\pySigma\sigma\modifiers.py", line 92, in apply
    r = self.modify(val)
        ^^^^^^^^^^^^^^^^
  File "C:\<redacted>\pySigma\sigma\modifiers.py", line 265, in modify
    return SigmaRegularExpression(val.original)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<string>", line 5, in __init__
  File "C:\<redacted>\pySigma\sigma\types.py", line 615, in __post_init__
    self.compile()
  File "C:\<redacted>\pySigma\sigma\types.py", line 629, in compile
    raise SigmaRegularExpressionError(
sigma.exceptions.SigmaRegularExpressionError: Regular expression 'something\$\' is invalid: bad escape (end of pattern) at position 11

Process finished with exit code 1
```

I'll keep digging to see if I can find the root cause.

**Comment by Res260** [comment]
When debugging, here is the difference between the original rule (bottom square, valid) and the `.to_dict()`'ed rule (top, invalid):

![image](https://github.com/user-attachments/assets/31addd25-04ef-4a52-a1eb-73bdfbf809ae)

**Comment by Res260** [comment]
I think I found the problem:

down the road of `SigmaRule.to_dict()`, `SigmaDetectionItem.to_plain()` is called and it takes the `original_value` property, which is of type `SigmaString` instead of `SigmaRegularExpression` which would probably escape properly.

<img width="690" alt="image" src="https://github.com/user-attachments/assets/b4dee598-d4d4-43a5-95f2-7b92d9cb49d7">


I'm not too sure how to fix that without breaking something. I'll give it a try.

**Comment by thomaspatzke** [comment]
That's possibly the reason because `original_value` contains the unmodified value, which is a `SigmaString` but this has different semantics than a regular expression.

**Comment by Res260** [comment]
Do you think there's a risk to change this to `value` instead?

**Comment by Res260** [comment]
Sorry for bump, this bug impacts us and I don’t mind fixing it, but I need a little bit more context to make a patch that doesn’t break something else :)

**Comment by thomaspatzke** [comment]
No problem! Sorry for the delays but that are busy times 😉

The missing backslash raises my suspicion if this is somehow related to issue #262. [This test](https://github.com/SigmaHQ/pySigma/blob/513e35285e7b4eb0044f9a3b439af833fde1b7db/tests/test_rule.py#L345-L349) goes into the same direction and doesn't fails. So it appears to be related to the slashes at the end.

In regards of determining if a patch breaks something, it is quite easy for pySigma. If the tests pass then nothing breaks. If something breaks regardless of the passing tests, then the tests were broken, not your patch 😁😉

**Comment by thomaspatzke** [comment]
I'be created a branch `issue-249` for this with some tests on this. Interesting is that [`SigmaRegularExpression.to_plain()` works well](https://github.com/SigmaHQ/pySigma/blob/7d47ae9e6324b77cb4cb80d656fb73a28c45ad2f/tests/test_types.py#L496) while the [same wrapped into a detection item lacks the last slash](https://github.com/SigmaHQ/pySigma/blob/7d47ae9e6324b77cb4cb80d656fb73a28c45ad2f/tests/test_rule.py#L352) which leads to the issue described by you.

**Comment by thomaspatzke** [comment]
Oops, I have to be more careful, both tests don't test exactly the same...

**Comment by thomaspatzke** [comment]
[This line](https://github.com/SigmaHQ/pySigma/blob/7d47ae9e6324b77cb4cb80d656fb73a28c45ad2f/sigma/rule.py#L372) seems to cause the drop of two slashes in `from_mapping()` which is the culprit, I believe.

**Comment by thomaspatzke** [comment]
@Res260 please check if the fix branch `issue-249` solves your problem.

**Comment by Res260** [comment]
Hey @thomaspatzke ! I just tested and it indeed fixes the issue :) 

I’ll wait for the next version of pySigma 🙏 Thanks for your help

**Comment by thomaspatzke** [comment]
Fixed in v0.11.11.
